src/metabase/query_processor/middleware/visualization_settings.clj

```diff
@@ -0,0 +1,72 @@
+(ns metabase.query-processor.middleware.visualization-settings
+  (:require [medley.core :as m]
+            [metabase.models.card :refer [Card]]
+            [metabase.query-processor.store :as qp.store]
+            [metabase.shared.models.visualization-settings :as mb.viz]
+            [toucan.db :as db]))
+
+(defn- normalize-field-settings
+  [id settings]
+  (mb.viz/db->norm-column-settings {(mb.viz/norm->db-column-ref {::mb.viz/field-id id}) settings}))
+
+(defn- update-card-viz-settings
+  "For each field, fetch its settings from the QP store, convert the settings into the normalized form
+  for visualization settings, and then merge in the card-level column settings."
+  [fields column-viz-settings]
+  (map (fn [[_ id _]]
+         (->> id
+              qp.store/field
+              :settings
+              (#(normalize-field-settings id %))
+              (#(m/deep-merge % (select-keys column-viz-settings [{::mb.viz/field-id id}])))))
+   fields))
+
+(defn- update-native-query-viz-settings
+  "For each column, select its viz settings from the column-viz-settings map by its name."
+  [columns column-viz-settings]
+  (map (fn [column] (-> column
+                        :name
+                        (#(select-keys column-viz-settings [{::mb.viz/column-name %}]))))
+       columns))
+
+(defn- get-viz-settings
+  "Pull viz settings from either the query map or the DB"
+  [query]
+  (if-let [card-id (-> query :info :card-id)]
+    ;; Saved card -> fetch viz settings from DB
+    (mb.viz/db->norm (db/select-one-field :visualization_settings Card :id card-id))
+    ;; Unsaved card -> viz settings passed in query
+    (-> query :viz-settings)))
+
+(defn- is-export-context?
+  [context]
+  (contains? #{:json-download :csv-download :xlsx-download} context))
```

Comment:
You could write this a little more generally by doing something like

```clj
(def export-contexts
  #{:json-download :csv-download :xlsx-download})
```

and then you can just do `(export-contexts context)` where you need it... I think that's probably defined somewhere else tho IIRC. Better if we can avoid duplicating the export contexts like so if we add a new one in the future we don't have to update a ton of different parts of the code.

Also, I think `export-context?` would be a little more idiomatic of a name for a predicate function like this -- e.g. Clojure core has `string?` and `number?` rather than `is-string?` and `is-number?`. Not that important tho

Reply:
Hmm, there's `export-formats` in `metabase.query-processor.streaming` but that returns `#{:csv :api :xlsx :json}`